core/src/peer_record.rs

```diff
@@ -0,0 +1,254 @@
+use crate::identity::Keypair;
+use crate::signed_envelope::SignedEnvelope;
+use crate::{peer_record_proto, signed_envelope, Multiaddr, PeerId};
+use std::convert::TryInto;
+use std::fmt;
+
+const PAYLOAD_TYPE: &str = "/libp2p/routing-state-record";
+const DOMAIN_SEP: &str = "libp2p-routing-state";
+
+// TODO: docs
+#[derive(Debug, PartialEq, Clone)]
+pub struct PeerRecord {
+    pub peer_id: PeerId,
+    pub seq: u64,
+    pub addresses: Vec<Multiaddr>,
+}
+
+impl PeerRecord {
+    // TODO: docs
+    pub fn into_protobuf_encoding(self) -> Vec<u8> {
+        use prost::Message;
+
+        let record = peer_record_proto::PeerRecord {
+            peer_id: self.peer_id.to_bytes(),
+            seq: self.seq,
+            addresses: self
+                .addresses
+                .into_iter()
+                .map(|m| peer_record_proto::peer_record::AddressInfo {
+                    multiaddr: m.to_vec(),
+                })
+                .collect(),
+        };
+
+        let mut buf = Vec::with_capacity(record.encoded_len());
+        record
+            .encode(&mut buf)
+            .expect("Vec<u8> provides capacity as needed");
+        buf
+    }
+
+    pub fn from_protobuf_encoding(bytes: &[u8]) -> Result<Self, DecodingError> {
+        use prost::Message;
+
+        let record = peer_record_proto::PeerRecord::decode(bytes)?;
+
+        Ok(Self {
+            peer_id: PeerId::from_bytes(&record.peer_id)?,
+            seq: record.seq,
+            addresses: record
+                .addresses
+                .into_iter()
+                .map(|a| a.multiaddr.try_into())
+                .collect::<Result<Vec<_>, _>>()?,
+        })
+    }
+
+    // TODO: docs
+    pub fn warp_in_envelope(self, key: Keypair) -> SignedEnvelope {
```

Comment:
As a first thought I would prefer passing the data to some component owning the keypair instead of passing the keypair to the data. The former would reduce the exposure of the keypair minimizing the possibilities for mistakes / attacks.

That said, the notion of peer records is still in an early phase, thus any abstractions that we come up with today are likely premature.

Reply:
Interesting idea but I don't quite see how it would work. Your identity keypair is constructed as the very first thing and usually only used during the construction of `Network`.

The way I see this being used is that the user just clones the KeyPair at that stage and also passes it to the Behaviour (like Rendezvous) for example. Such a behaviour can then further use it to construct `SignedEnvelopes` or other things.